File: src/render.ts
/**
 * Build the HUD's DOM from an analysed document.
 *
 * Deliberately mirrors `rtlens_core::render` element for element and class for class, so
 * the offline harness the engine is tested against and the shipping window are the same
 * layout. Any change here needs the matching change there.
 */

export type Dir = "ltr" | "rtl";
export type Segment = { kind: "text" | "code"; text: string };

export type Cell = { dir: Dir; align: Dir; segments: Segment[] };

/**
 * Table membership. Table-wide facts repeat on every row so the wire format stays a flat
 * list of lines; consecutive rows sharing an `id` are one table.
 */
export type TableRow = {
  id: number;
  columns: number;
  dir: Dir;
  head: boolean;
  framed: boolean;
  cells: Cell[];
};

export type Line = {
  lead: string;
  tail: string;
  boxed: boolean;
  code: boolean;
  dir: Dir;
  segments: Segment[];
  table?: TableRow;
};

export type CaptureSource = "selection" | "clipboardFallback" | "empty";

export type Doc = {
  lines: Line[];
  source: CaptureSource;
  diffMode: boolean;
};

function isBlank(line: Line): boolean {
  return (
    line.lead.trim() === "" &&
    line.tail.trim() === "" &&
    line.segments.every((s) => s.text.trim() === "")
  );
}

/** Keep rendered text selectable inside the HUD's deep window-drag region. */
function preserveTextSelection(element: HTMLElement): void {
  element.setAttribute("data-tauri-drag-region", "false");
}

function gutter(text: string, tail = false): HTMLElement {
  const span = document.createElement("span");
  span.className = tail ? "rtl-gutter rtl-gutter--tail" : "rtl-gutter";
  span.textContent = text;
  preserveTextSelection(span);
  return span;
}

function appendSegments(target: HTMLElement, segments: Segment[]): void {
  for (const segment of segments) {
    if (segment.kind === "code") {
      const code = document.createElement("span");
      code.className = "rtl-code";
      code.setAttribute("dir", "ltr");
      code.textContent = segment.text;
      target.append(code);
    } else {
      target.append(document.createTextNode(segment.text));
    }
  }
}

/**
 * A recovered table, as one grid. Every row of a table has to live in the same grid or the
 * browser has no way to give them shared column widths.
 */
function tableElement(rows: TableRow[]): HTMLElement {
  const table = document.createElement("div");
  table.className = rows[0].framed ? "rtl-table rtl-table--framed" : "rtl-table";
  // The table's direction is its column order: a Persian table starts at the right.
  table.setAttribute("dir", rows[0].dir);
  table.style.setProperty("--rtl-cols", String(rows[0].columns));
  for (const row of rows) {
    const rowEl = document.createElement("div");
    rowEl.className = row.head ? "rtl-row rtl-row--head" : "rtl-row";
    for (const cell of row.cells) {
      const cellEl = document.createElement("div");
      // `dir` is the cell's own direction; the class is its column's, which is what the
      // text aligns to so a column reads as one column.
      cellEl.className = `rtl-cell rtl-cell--${cell.align}`;
      cellEl.setAttribute("dir", cell.dir);
      preserveTextSelection(cellEl);
      appendSegments(cellEl, cell.segments);
      rowEl.append(cellEl);
    }
    table.append(rowEl);
  }
  return table;
}

function lineElement(line: Line): HTMLElement {
  const el = document.createElement("div");
  el.classList.add("rtl-line");
  if (line.code) el.classList.add("rtl-line--code");
  if (isBlank(line)) el.classList.add("rtl-line--blank");
  // A box frame is structural and stays physical; every unboxed RTL line belongs on the
  // reading-start side and so flips with the line. The presence of a lead is unrelated
  // to the line's writing direction.
  if (!line.boxed && line.dir === "rtl") {
    el.classList.add("rtl-line--rtl");
  }

  if (line.lead !== "") el.append(gutter(line.lead));

  const content = document.createElement("span");
  content.className = "rtl-content";
  content.setAttribute("dir", line.dir);
  preserveTextSelection(content);
  appendSegments(content, line.segments);
  el.append(content);

  if (line.tail !== "") el.append(gutter(line.tail, true));
  return el;
}

export function renderDoc(target: HTMLElement, doc: Doc): void {
  const fragment = document.createDocumentFragment();
  let i = 0;
  while (i < doc.lines.length) {
    const first = doc.lines[i].table;
    if (first) {
      const rows: TableRow[] = [first];
      let end = i + 1;
      for (; end < doc.lines.length; end += 1) {
        const next = doc.lines[end].table;
        if (!next || next.id !== first.id) break;
        rows.push(next);
      }
      fragment.append(tableElement(rows));
      i = end;
    } else {
      fragment.append(lineElement(doc.lines[i]));
      i += 1;
    }
  }
  target.replaceChildren(fragment);
}

File: crates/rtlens-core/src/render.rs
//! Render a [`Doc`] to HTML.
//!
//! The shipping HUD builds the same DOM in TypeScript; this exists so the engine can be
//! inspected visually offline, and so snapshot tests assert on real markup rather than on
//! a debug format that could drift from what users see.

use crate::model::{Chunk, Doc, Line, Segment, TableRow};

/// The canonical layout rules, shared verbatim with the frontend.
pub const LAYOUT_CSS: &str = include_str!("../../../src/styles/layout.css");

fn escape(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        match c {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            _ => out.push(c),
        }
    }
    out
}

fn segments_html(segments: &[Segment]) -> String {
    let mut html = String::new();
    for seg in segments {
        match seg {
            Segment::Text(t) => html.push_str(&escape(t)),
            Segment::Code(t) => {
                html.push_str(&format!("<span class=\"rtl-code\" dir=\"ltr\">{}</span>", escape(t)))
            }
        }
    }
    html
}

/// A recovered table, as a grid.
///
/// The columns are `auto`-sized by the browser rather than by the widths the terminal
/// painted: those widths were chosen for a monospace grid, and the HUD is proportional.
/// What has to survive is the column *structure*, not its former pixel geometry.
fn table_html(rows: &[&TableRow]) -> String {
    let first = rows[0];
    let mut classes = String::from("rtl-table");
    if first.framed {
        classes.push_str(" rtl-table--framed");
    }
    let mut html = format!(
        "<div class=\"{}\" dir=\"{}\" style=\"--rtl-cols:{}\">",
        classes,
        first.dir.as_attr(),
        first.columns
    );
    for row in rows {
        html.push_str(if row.head {
            "<div class=\"rtl-row rtl-row--head\">"
        } else {
            "<div class=\"rtl-row\">"
        });
        for cell in &row.cells {
            html.push_str(&format!(
                "<div class=\"rtl-cell rtl-cell--{}\" dir=\"{}\">{}</div>",
                cell.align.as_attr(),
                cell.dir.as_attr(),
                segments_html(&cell.segments)
            ));
        }
        html.push_str("</div>");
    }
    html.push_str("</div>");
    html
}

fn line_html(line: &Line) -> String {
    let mut classes = vec!["rtl-line".to_string()];
    if line.code {
        classes.push("rtl-line--code".into());
    }
    if line.is_blank() {
        classes.push("rtl-line--blank".into());
    }
    // A frame stays physical; a marker or indent follows the reading direction.
    if !line.boxed && line.dir == crate::Dir::Rtl && !line.lead.is_empty() {
        classes.push("rtl-line--rtl".into());
    }

    let mut html = format!("<div class=\"{}\">", classes.join(" "));
    if !line.lead.is_empty() {
        html.push_str(&format!("<span class=\"rtl-gutter\">{}</span>", escape(&line.lead)));
    }
    html.push_str(&format!("<span class=\"rtl-content\" dir=\"{}\">", line.dir.as_attr()));
    html.push_str(&segments_html(&line.segments));
    html.push_str("</span>");
    if !line.tail.is_empty() {
        html.push_str(&format!("<span class=\"rtl-gutter rtl-gutter--tail\">{}</span>", escape(&line.tail)));
    }
    html.push_str("</div>");
    html
}

/// The document body, without any wrapper chrome.
pub fn fragment(doc: &Doc) -> String {
    let parts: Vec<String> = doc
        .chunks()
        .into_iter()
        .map(|chunk| match chunk {
            Chunk::Line(line) => line_html(line),
            Chunk::Table(rows) => table_html(&rows),
        })
        .collect();
    format!("<div class=\"rtl-doc\">{}</div>", parts.join("\n"))
}

/// A self-contained page for offline inspection.
pub fn standalone(doc: &Doc, title: &str) -> String {
    format!(
        "<!doctype html>\n<html><head><meta charset=\"utf-8\">\n<title>{}</title>\n\
         <style>\nbody {{ margin: 0; padding: 24px; background: #14161a; color: #e6e6e6; }}\n\
         {}\n</style></head>\n<body>\n{}\n</body></html>\n",
        escape(title),
        LAYOUT_CSS,
        fragment(doc)
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{process, CaptureSource, Options};

    fn html(input: &str) -> String {
        fragment(&process(input, &Options::default(), CaptureSource::Selection))
    }

    #[test]
    fn sets_dir_per_line() {
        let out = html("سلام\nhello");
        assert!(out.contains(r#"<span class="rtl-content" dir="rtl">"#));
        assert!(out.contains(r#"<span class="rtl-content" dir="ltr">"#));
    }

    #[test]
    fn code_runs_are_ltr_isolated() {
        let out = html("فایل /src/main.rs را ببین");
        assert!(out.contains(r#"<span class="rtl-code" dir="ltr">/src/main.rs</span>"#));
    }

    #[test]
    fn box_frame_lands_in_gutters_not_content() {
        let out = html("│ سلام │");
        assert!(out.contains(r#"<span class="rtl-gutter">│ </span>"#));
        assert!(out.contains("rtl-gutter--tail"));
        // The frame must never end up inside the directional content span.
        let content = out.split(r#"class="rtl-content""#).nth(1).unwrap();
        let content = content.split("</span>").next().unwrap();
        assert!(!content.contains('│'));
    }

    #[test]
    fn rtl_bullet_line_flips_but_boxed_line_does_not() {
        assert!(html("- سلام").contains("rtl-line--rtl"));
        assert!(!html("│ سلام │").contains("rtl-line--rtl"));
    }

    #[test]
    fn escapes_markup_in_content() {
        let out = html("<script>alert(1)</script>");
        assert!(!out.contains("<script>"));
        assert!(out.contains("&lt;script&gt;"));
    }

    #[test]
    fn standalone_page_embeds_the_shared_css() {
        let d = process("سلام", &Options::default(), CaptureSource::Selection);
        let page = standalone(&d, "t");
        assert!(page.contains("unicode-bidi: isolate"));
        assert!(page.starts_with("<!doctype html>"));
    }
}
